### ISSUE TITLE
Redfine <LENGTH> as 1 byte, unless 0 then follow with large <LENGTH>

### ISSUE THREAD

**Issue by MikeFair** [thebuzzmedia/universal-binary-json#66]
I recommend/propose changing <LENGTH> to be defined as:
< LENGTH > = (< U > != 0) ? < U > : < TYPE >< VALUE >

(I'm not sure if my syntax is clear/correct there, but what it says is:
Whenever a LENGTH is called for, it assumes the default will be <= 255 (1-unsinged byte; a uint8).
If >255 length is needed, this byte is set to 0, then the usual < LENGTH > specification follows.

Currently < LENGTH > always costs an extra byte to declare the type and requires the decoder to interpret the type character before it can tell how many bytes the next integer will be costing decoding time/adding a little complexity, even it's going to be a [U] anyway.  

This recommendation reduces the size of the < LENGTH > in all the smaller cases (which I think is easy to see will be more common than large cases - just think about how many strings <=255 there are) and it makes decoding faster because the decoder can assume the length will be defined in just the next byte for the common case, then only if that byte is a zero does it need to interpret the next item for length.

**Comment by thebuzzmedia** [comment]
@MikeFair I like the spirit behind this optimization, but I'm wondering how you could tell the difference between `[S][U][3][bob]` and `[S][85][<85 chars>]` after parsing the 'S' then the decimal value of '85' (which is ascii value of 'U')

**Comment by MikeFair** [comment]
@thebuzzmedia 

This spec changes it so you can't do use [S][U][3] anymore, it would be just [S][3]
If you needed something longer than 255 (like 1024) it would be: [S][0][I][1024]
It introduces the cost of an extra byte in the "long" case, in favor of optimizing the short case.

**Comment by MikeFair** [comment]
@thebuzzmedia 

If you're looking for backward compatibility, aside from just excluding the special cases of 85 and the other integer types (which I think is a bit awkward), I don't think you can, which is why this spec makes a difference now as it is introducing a backward incompatible change.

**Comment by thebuzzmedia** [comment]
@MikeFair right-o... it would also break the intended structure of the spec (type-length-data for everything) for savings that, to your point, I do think would add up, but I don't necessarily think are worth the cost.

But I want to hear thoughts from others... if everyone comes in here and says "Yes, this is how it SHOULD have been and we will never forgive you for that", then I would consider the breaking change especially since we haven't done a 1.0 yet.

**Comment by MikeFair** [comment]
@thebuzzmedia

This completely preserves the TLV stucture it is simply redefining how L is
defined.  :)
The idea is anywhere L gets used would behave like this.

**Comment by Miosss** [comment]
For parsers, this optimization means only one more if:
```
if(peekByte() != 0)
      count = getByte();
else
    ; // old code
```
Similar one-liner for decoder.

Is it optimization? Yes, a small one.
Does it make encoders/decoders more complex? Yes, a little.
Does the specification change? Yes, it becomes a little less structured.

What is the effect? I do not know, probably gains and losses are equal. But still it is only 1 byte for a value, in most cases. It would be nice to have some training data to check, how does it look in real usage.

**Comment by MikeFair** [comment]
I think a better way to code this to take better advantage of the technique would be this instead:
<pre>
    count = getByte();
    if(count != 0) return count;
    ; // old code
</pre>

When using the peekByte() approach, the code is doing the work to translate the byte to a count, checking its value, then translating that byte again. 
Through knowing that there is always a first Byte count in the LENGTH, call getByte() optimistically, as long as it's not 0 (the extremely common case) it can immediately return it; if it is [0] then that first Byte needs to be consumed anyway to get the Byte after that which gives you the length of the int that follows.  (In the original code posted above, the // old code; would first encounter the [0] again.)

Aside from just code savings, which I think is saving just an extra data fetch call, the main advantage of the technique is the space because it is eliminating 1 Byte for every [U] sized length.

>  It would be nice to have some training data to check, how does it look in real usage.

I agree some real usage training statistics would be ideal.

I don't have anything offhand, but some Google searching comes up with some good datasets for a small idea on exactly how much space saving opportunity this idea creates:
* Data.gov appears to supply JSON: (e.g. http://catalog.data.gov/dataset/enterprise-data-inventory)
* http://www.yelp.com/dataset_challenge/ (Need to sign up for a free API account)
* http://labrosa.ee.columbia.edu/millionsong/lastfm (Look for zip files: SUBSET -- TRAIN SET --- TEST SET )
* http://2014.recsyschallenge.com/dataset/ (Movies in CSV w/ a field of JSON tweetdata)
* OpenStreetMap; I couldn't find a direct GeoJSON download of the OSM database but found a number of converters to GeoJSON from OSM; the original OSM can be found here: http://wiki.openstreetmap.org/wiki/Downloading_data

Just quickly looking at some records in these samples, the number of "lengths" <=255 that exist is mind blowing (way more than I originally suspected, and I was already thinking "a lot").  I was most surprised to see even the GeoJSON dataset had many arrays and strings that would fit this constraint (and of course several that wouldn't as well).  As expected, almost all of them are pretty heavy on the arrays of objects (meaning lots of short (<256) field name strings).

@thebuzzmedia 
Here's a thought, can some ideas/issues be flagged for "inclusion in the next backward incompatible draft"; that way in situations like this, where there's clearly a benefit, but it doesn't seem to warrant a "break" by itself, it can be flagged for inclusion into the next version of the spec that does.

**Comment by Miosss** [comment]
> I think a better way to code this to take better advantage of the technique would be this instead:

Gosh, that was just quick pseudocode...


> Aside from just code savings, which I think is saving just an extra data fetch call, the main advantage of the technique is the space because it is eliminating 1 Byte for every [U] sized length.

The space is main thing we try to optimize.


In fact, length only affects strings, I recall now. If you exclude all container optimizations, and say that length of strings would be defined as one byte, and if that is zero we go to the old definition - then I think I agree.

Provided that this issue concerns only string-length, I think that @MikeFair proposal has fair sense. Probably around 95% of all strings send in JSON are shorter than 255 bytes, thus we gain one extra byte every so often. 
Maybe that 1B is not much, but even something  like:

```
[S][15][hello]
```

is easier to understand than

```
[S][i][15][hello]
```

**Comment by kxepal** [comment]
Instead of breaking TLV and adding more branches to the parser why not to introduce separate tag which issues only one-byte-length sized strings? Say `[s][6][hello!]` while logic of `[S]` will remains consistent.

**Comment by Miosss** [comment]
@kxepal 
Nice idea, but then we would have two markers conveying the same datatype. Is it reasonable?

But if we introduce ```[s]```, shall we enforce parsers to use ```[s]``` for all string shorter than 255 bytes? For example do we exclude ```[i] and [U]``` from LENGTH for strings? It makes mess in specification (we would have to introduce something like: any-integer-type-longer-than-8bits). If not, that we could write both:

```
[S][i][15][hello]
```

and

```
[s][15][hello]
```

which is a bit confusing too.

**Comment by kxepal** [comment]
@Miosss we already have 8 of them to handle one JSON datatype named as "number" (: So it depends what to call "datatype". Short strings `[s]` plays the role as `[U]` or `[i]` plays for `[I]`.

> shall we enforce parsers to use [s] for all string shorter than 255 bytes?

No, we shall not since `[s]` and `[S][i]` and `[S][U]`  are interchangeable. Encoders may or may not use such optimizations, but decoders shouldn't care much about since the result will be the same kind of string.

**Comment by Miosss** [comment]
@kxepal 
Yes, I know, but numerics can have other form in some languages (ints, longs, short ints, floats, etc.) so in some sense, there is difference between them all.
In this case, actual data (the string) is the same, only its description changes.

This is of course splitting a hair into four.
I admit, that introducing ```[s]``` gives the advantages of original proposition, but leaving specification almost as clear as know. Thats a +

About the enforcing.
I agree that decoders should be ready for decoding ```[S][i]``` either way. But I insist that in such case ```[s]``` is enforced, maybe as a "developer guideline". There is no sense in writing ```[S][i]``` from when we will have ```[s]```.

**Comment by kxepal** [comment]
@Miosss you're right and some languages provides even more types for numbers as we have (:

About enforcing your insist is completely correct in order of best practices as like we should encourage developers to use UBJSON format effectively and discourage the non-efficient ways. But we cannot tell devs to reject such kind of data during decoding since it's still a valid UBJSON data, just baked in not effective way.

P.S. Btw, using `[S][i]` is incorrect too, since `[i]` is signed byte while `[U]` is unsigned - string length cannot be negative so `[U]` gives bigger capacity for the same size (;

**Comment by Miosss** [comment]
> But we cannot tell devs to reject such kind of data during decoding since it's still a valid UBJSON data, just baked in not effective way.

I agree. So we should enforce correct encoding behaviour, while maintaining full decoder capabilities, even in those 2 narrow, unoptimal cases where we do not make spec more complex: in writing ```[S][i] and [S][U]```.

In such case, for now, I would vote for ```[s]``` instead of ```[0]```.

**Comment by Miosss** [comment]
@kxepal 
Ok, there is a problem.

```[S]``` is optimized out for object-keys. We could of course change default type to ```[s]``` but it will not be backwards compatible. And there must be option for longer keys. What will we write then, ```[0]``` ? 


Keys longer then 255 are probably rare sa hell, but there is no limit and we cannot create one.

**Comment by kxepal** [comment]
@Miosss great point (: But oblivious solution is to not apply `[s]` optimization for object keys or revert object keys optimization in order to specify keys type explicitly - for 80% cases keys aren't longer than 255 chars (even unicode ones) so there will be no loss and no gain, but we'll solve this collision gracefully (and return ability to have one-char keys by using `[C]` which wasn't removed).

So I think all the problems are solves, aren't they? (:

**Comment by Miosss** [comment]
@kxepal 
It solves the problem, but keys are approximately 60-70% of strings in JSONs. More than that, it is more likely to find short string in object keys than in values.
Using this solution we nuke almost all pros of this approach and get nothing in return.
Using something like ```[0]``` for key longer than 255 bytes is sooooo inconsistent in the spec, but I believe it would happen really occasionaly. I do not find it as very good solution though.

Btw. I am absolutely distressed with the fact, that @thebuzzmedia  left ```[C]``` after agreeing on its removal (:

**Comment by kxepal** [comment]
@Miosss keys are ALWAYS strings in JSON (: or this is invalid JSON. So there is no issue at all.

I also didn't found `[C]` very useful, but some mongodb folks may love it after all (:

**Comment by Miosss** [comment]
@kxepal 
Yes I know. I meant, that about 70% of strings used in JSON messages are those that form keys (I didnt write that 70% of keys are strings :) ).

**Comment by kxepal** [comment]
@Miosss sorry, completely misread you ((: Will try again...

Well, indeed, there no profit happens with the only exception that spec becomes a little bit more cleaner: we remove using implicit type from objects while we give global optimization for short strings without breaking the TLV. I think it's worth to have such in not terms to receive some big optimization, but to clean things up. For instance [this example](https://github.com/thebuzzmedia/universal-binary-json-java/blob/master/src/test/resources/org/ubjson/TwitterTimeline.formatted.json) will be optimized additionally by 24 bytes. Not much, but still a something.

In anycase some fast compressor like snappy will compact data much better than we could do by playing with tags and values (:

**Comment by Miosss** [comment]
@kxepal
But removing ```[S]``` from each key declaration is just sooo obvious! It is almost natural. + it gives one extra byte out of the box.

About compression.
I think that we should focus on gzip (zlib) and deflate, because they are widespread in HTTP, Snappy seems to be also of concern but on second place.
But more generally, what we do in UBJSON is creating domain-specific optimizations. We optimize things that we now we can in spite of JSON/UBJSON specification. What general compression algorithms do, is to throw away any additional information and compress raw stream of bytes. By dictionaries or by entropy, they perform very well. But our optimizations here can remove some things, that they would not be able to.

So, UBJSON + gzip will still be more effective then JSON + gzip (you can probably find some ill examples though).

**Comment by kxepal** [comment]
> But removing [S] from each key declaration is just sooo obvious! It is almost natural. + it gives one extra byte out of the box.

Yes, but it also sooo hacky from the point of TLV and overall UBJSON due to implicit assumptions. Clean and explicit rules leads to clean and simple implementation. From the point of implementation it leads to inability to have a single function parse that just emits tag-length-value structure for further processing without additional I/O operations....but here are goes irrelevant details (:

**Comment by Miosss** [comment]
Yes you are right.

So we have 3 options so far, I think:
- make ```[U]``` the default length-specifier; if ```[0]``` encountered where length was expected, fall back to full type-value declaration
- introduce ```[s]``` which is equivalent in semantics to ```[S][U]```; in addition to that, we have to reinsert full specification of keys in object, so each key is not of type ```[S]``` by default; possiible types will be: ```[S] [s] and [C]``` (maybe we could drop [C] at last...)
- only change the meaning of ```[S]``` which will now be by default interpreted as ```[S][U]```; if expected length-byte shall be equal to 0, we fall back as in first point - full length-type specification

The first is original @MikeFair proposal.
The second is @kxepal 's modification.
And the third is a mixture of first and second, without changing behaviour of objects much.

Btw. there still is ```[H]``` (high-precision) which is, in the matter of fact, identicall to string. Must we then introduce ```[h]``` to be consistent?

**Comment by MikeFair** [comment]
A Length is used in repeating headers/optimized containers too (not only
strings).

What I'm following is how tgis breaks TLV...
Why is the L in TLV is being treated like it needs/should be a TLV itself.

TLV is already regularly broken because L is ommitted for most types; and
so is already quite special.  It's not written [U][1][<value>] which would
be the correct TLV description.

What about this; create this as a separate version of Lengh called [l] and
keep the existing one as [L] then let the spec use the one it considers
best...  [S] uses [L], [s] uses [l], and future types that need lengths use
[L] or [l] as they see fit.

(As for the lengths of bytes greater 8 problem, keep in mind that 6 bytes
can count up to 256TB; I don't even know what they call 8 bytes worth of
storage.)

**Comment by Miosss** [comment]
@MikeFair 
What are you talking about? TLV is just a name, what do you mean by [L] and [l] - this is nothing in the spec!

We cannot also discuss N-ways to optimize containers and on the other hand discuss ommiting some length-specifiers in strings. Containers are suuuch complicated and we have sooo many different ideas that this length from this issue has nothing to do with it.

And where from 8Bytes? 8 bits == 1 byte is what you proposed, so 255 values of length.

**Comment by MikeFair** [comment]
While I thought it was a clever idea to use a new type to define the
different ways to get a count; when I looked at the three options put
forth; every one of them breaks backwards compatibility in some way anyway.

The good thing that came from [s] was preserving the backward compatibility
of [S], which ends up eliminating the benefit for object keys (where the
main benefit of the idea will likely come from).

You could also create a new short version of {, and [h] and optimized
containers but let's take a step back and look at this a second.

Since we can't save both [S] and [{] accept this as a backwards
incompatible change.

Unless someone can see soke clever way to detect which layout an object is
using during a decoding, all options presented are going to break something.

Isn't it better to break it cleanly and just redefine length everywhere for
all things in a consistent way then try and patch something that cannot be
saved.

Or say "We're not going to do it; or shouldn't do it until we have other
backward breaking changes to go with it".

That said, here's an option (I was planning on postijg this as another
issue anyway)
4) Make a UBJ header field at the start of the stream (maybe encapsulated
between <> or the negative equivalent of {}) and put a format version
number in it.  A header is likely to be useful for other things too (but
obviously keep it _small_).

Let new parsers detect which rules to follow (or to error out on a version
they can't decode) and in the absence of a header, fall back to exisiting
behavior.
Old parser are just out of luck, but they were already out of luck anyway.

---

>
> And where from 8Bytes?

Forget I said anything about that, I completely misread the comments (I
somehow read needing a length longer than 8 bytes - which is nothing anyone
has said).

> What are you talking about? TLV is just a name, what do you mean by [L]
and [l] - this is nothing in the spec!

I keep seeing this phrase:

> Instead of breaking TLV

Doesn't TLV refer to [Type] [Length] [Value]?
That's the common meaning I'm used to seeing...

T   L     V
[S] [5] [hello]

And although it's called [type] [length] [data] on this page; here's the
spec source for it:
http://ubjson.org/#data_format

So how does this idea break TLV, it doesn't.  All things are still a TLV;
it just breaks how L in that TLV is read in.  It does break the existing
reading rule of counts and lengths; and replaces it with a new, clear rule.

What I was saying about optimized containers is that they also use counts
(or lengths).

Which means this issue can also apply to them (and the intention is for all
counts/lengths to work this way).
http://ubjson.org/type-reference/container-types/#optimized-format

**Comment by kxepal** [comment]
@Miosss 

> Btw. there still is [H] (high-precision) which is, in the matter of fact, identicall to string. Must we then introduce [h] to be consistent?

Another good point! Feelings are so good as in Draft-8 days when we had both `[s]` and `[h]` ((: And `[h]` makes much more sense than `[s]` since this is the only type that guarantee floating point precision and it's widely have to be used for shot numbers as well.

@MikeFair 

I would avoid to add another brackets just in order to apply keys optimization for the objects. There is much more useful case for them is to simplify how we define containers since all these headers, typedefs and the other `#@` markers inside are overloads too much a single container tags.

> So how does this idea break TLV, it doesn't.

Let's define a grammar of UBJSON L thing:
INTEGER = 0x55, 0x69, 0x49, 0x6C, 0x4c
BINARY = <any>
L = <INTEGER><BINARY>

It's breaking in the way that you adds a branch when `0x00` is acceptable L definition: 

UNSIGNED-BYTE= 0x00-0xFF
L = <UNSIGNED-BYTE>, 0x00<INTEGER><BINARY>

Which means that L could be defined in two different ways using different kind of logic to handle it value. Also, have you find a way to define zero length strings with your proposal? Having `[S][0][U][0]` is a VERY strange notation for empty strings.

How this breaks whole TLV? Easily with help of domino effect:

- T is required, L is optional, V is optional (see site and below)
- Currently L is defined recursively through TLV definition of integer types
- Allowing L to be any byte means that T isn't strictly defined anymore by some given list
- The 0x00<INTEGER><BINARY> assumes to be TLV structure as well, but 0x00 isn't a valid T marker
- BOOM!

Collateral damage:  type definition have to be changed in the way to be ugly (from the site):  

```
type
        A 1-byte ASCII char used to indicate the type of the data following it.
length (OPTIONAL)
        A positive, integer numeric type (int8, uint8, int16, int32, int64) specifying the length of the following data payload leaded by null byte or just a unsigned byte only for string types
data (OPTIONAL)
        A run of bytes representing the actual binary data for this type of value.
```

Sounds really awkward, right?

**Comment by Miosss** [comment]
@kxepal 
> Another good point! Feelings are so good as in Draft-8 days when we had both [s] and [h] ((: And [h] makes much more sense than [s] since this is the only type that guarantee floating point precision and it's widely have to be used for shot numbers as well.

Did ```[s]``` and ```[h]``` mean in Draft8 this, what we are talkign about here? If so, why they were dropped? (Maybe there is something important I'm missing)

@MikeFair 
> Isn't it better to break it cleanly and just redefine length everywhere for
all things in a consistent way then try and patch something that cannot be
saved.

But **length** as we discuss here is only valid in strings and high-precision numbers. All other types containt bultin length value (int32 -> 4 bytes, etc.).
Strongly Typec Containers (current spec), Steve's NDs and kxepal's **typespec**, all of them have their own ideas, scpecifications, markers, etc. Do not mix simple string length with those, hugely optimizing constructs, which are yet to be included / discarded.

**Comment by kxepal** [comment]
> Did [s] and [h] mean in Draft8 this, what we are talkign about here? If so, why they were dropped? (Maybe there is something important I'm missing)

Yes, they were used for one byte length values while `[S]` and `[H]` were used for 4-bytes ones. They were dropped because length specification became more flexible (nowdays you may have strings 2 and 8 bytes sized as well) and there were no reason to keep them after all. Today, it seems we have found a way to get them back (:

**Comment by kxepal** [comment]
cc @thebuzzmedia ^^^

**Comment by thebuzzmedia** [comment]
**Disclaimer** I skimmed the last 10 replies, so if I missed a key point let me know...

* I do not want to re-introduce special-case markers (S/s, H/h, U/u) -- we took them out for consistency/simplicity sake, I don't want to add them back.
* Adding branching logic to the parsing of a length (1 `if`) is no different than adding it directly in the stream parser that checks for `S` or `s` -- so net-net machine code executed during parsing will likely be close to identical, so I don't see that as a big argument _against_ making the change that @MikeFair proposes.
* I AGREE that a huge amount of data passed in most JSON APIs are specifically strings - and as was mentioned above, "95%" of them are < 255 in length, so I very much like @MikeFair proposal for that reason.
* In cases of data > 255, parsing the entire string is going to be so much more expensive that introducing the extra byte/if-check will be a trivial amount of microseconds added to the processing -- to that point, I feel like Mike's proposal here is closing a gap on something I missed before -- he is optimizing for the "95%" case at the expense of 1-byte unoptimization for the "5%" case, which is exactly the right tradeoff to make -- I was previously so focused on 'unification' in Draft 9/10 that I ignored that.
* @MikeFair to your specific question about a certain kind of label for breaking/non-breaking changes -- I am still entertaining ALL kinds of changes, breaking included, pre-1.0 -- so no need for this yet.

As-proposed, are you guys mostly in favor of @MikeFair proposal to add it to Draft 12? If not, let me know why (or reference a comment above if I missed a strong argument against when I skimmed - sorry if I did!)

**Comment by kxepal** [comment]
@thebuzzmedia you really miss the last 10 replies (: If they won't convince you, let me know to try again.

**Comment by thebuzzmedia** [comment]
Uh oh, my apologies, reading more closely...

**Comment by MikeFair** [comment]
@kxepal 

> Also, have you find a way to define zero length strings with your proposal? Having  [S][0][U][0]  is a VERY strange notation for empty strings.

Excellent point!!  I totally overlooked that case!

How about using [s] to mean the empty string?  (I've got a proper fix, but think about it for a second)
And I do mean explicitly just [s] (like [T] and [F]) and not introduce [s] to mean "short string".

There's also the proper fix, in which I've gone ahead and modified the proposal to use 0xFF instead of 0x00 for the long case.
Which only means only encoding strings exactly 255 characters long get removed in favor of including empty strings.

**Comment by MikeFair** [comment]
@Miosss 
> But length as we discuss here is only valid in strings and high-precision numbers.

I disagree on that.  This style of Length ought to include more than just Strings and High-Precision numbers.  The existing Object Key, existing Optimized Container format, and upcoming proposals  ought to also be considered. e.g. a section header/footer; a checkpoint record (which has been considered); or other optimized containers (regardless of method chosen) are all very likely to need/include a count.  

When future proposals are being put forth, and they include counts in their examples, which format would you like to see them using?  

I think they should be typing [12] instead of [U][12] when a count of 12 is called for, and (using the newly revised proposal) [255][I][1024] when a length/count of 1024 is called for.

If this style of Count is considered the exception, and we invented a couple type characters for it, then it will properly be treated like the exception; if we make this "the rule" then in future proposal the "long form" of count (where you always explicitly specify the int type) becomes the exception.

**Comment by thebuzzmedia** [comment]
@kxepal Re-read everything, my opinion is largely unchanged, but I do see some clarification needed.

ALL, I am interpreting this entire thread to be @MikeFair suggesting we change the definition of _length_ (`L` in `TLV`) to be:
```
L = (0-254) | (255 && (i | U | I | l | L))
```

This would to apply to every length we have defined in the spec -- container lengths, string lengths, high-precision number lengths...

Examples:

```
[S][0] -- empty string
[S][3][foo] -- 3 char string
[S][255][I][4096][...] -- 4k string
[H][0] - empty high prec number
[H][8][1.234567]

[[][#][0] - 0 element array header
[[][#][3] - 3 element array header
[[][#][255][I][1024] - 1024 element array header

[{]
  [3][foo][S][3][bar] - "foo":"bar"
  [255][I][874][...874 char label...][S][3][wow] - "<874 char label>":"wow"
[}]
```

etc. etc.

This is the only change I want to focus on with this particular issue - I'm still against the re-introduction of lowercased type markers.

As I mentioned above, I am in support of this change because, while a little odd, I think it optimizes for 90%+ of the payloads (Especially String) that will be transferred in UBJSON and in the case where payloads are bigger than 254, I think the additional bytes are trivial parsing/logic alongside the bigger data payloads that need to be parsed and shuffled around.

Thoughts?

**Comment by kxepal** [comment]
@thebuzzmedia 

Ok, let's compare both solutions:

```
[S][0] vs [s][0] -- empty string
[S][3][foo] vs [s][3][foo] -- 3 char string
[S][255][I][255][...] vs [s][255][...] -- 255 chars string (oops!)
[S][255][I][4096][...] vs [S][I][4096][...] -- 4k string
[H][0] vs [h][0] - empty high prec number
[H][8][1.234567] vs [h][8][1.234567]
```

So far no difference with exception that we don't need to have magic `[255]` if string is longer than 254 bytes and we don't need to pay additional byte for strings with length exactly 255 bytes. For every string  longer than 254 bytes we don't have to pay for one additional byte.

```
[{]
  [3][foo][S][3][bar] vs [s][3][foo][S][3][bar] - "foo":"bar"
  [255][I][874][...874 char label...][S][3][wow] vs [S][I][874][...][S][3][wow] - "<874 char label>":"wow"
[}]
```

So far again no much difference, but again one magic marker lesser. Bonus:

```
[{]
  [C][a][S][3][bar] - "a":"bar"
[}]
```

Since `[C]` wasn't removed, it's unclear way it couldn't be used for object keys. Now this problem is solved.

As for the containers case:

```
[[][#][0] - 0 element array header
[[][#][3] - 3 element array header
[[][#][255][I][1024] - 1024 element array header
```

This may attract at the first sight, but such kind of optimization doesn't gives much space saving since container length defines only once. I don't think optimization at this place plays any significant role.

From point of implementation `[255]` solution causes to change all the parses in the way to handle it. Introducing a new `[s]` and `[h]` applies transparently without parser change since all what you need is to add new tags handlers.

Resume: the @MikeFair proposal is really good, but the proposed implementation while optimizes 90% of cases, introduces unwanted overhead for the other 10%. There is an alternative solution that doesn't have such flaw. Additionally,  it accidentally causes specification cleanup in order to remove implicit type tags from object keys with no loss in comparison with @MikeFair proposal and is able to make one-char keys legal which allows to save one more byte for them. The one-char keys are very popular for mongodb community since this is the way how they fight for the memory and space there. They are also makes a sense in scientific domain.

From point of specification: does it needs to be clean, consistent and fluent with as less as possible edge cases or not? Simplicity is a key feature of UBJSON and branches for magic bytes were never made things simple.

**Comment by MikeFair** [comment]
@kxepal 

I think you overlooked something in those comparisons.
The [s] approach yields no gains in the case of typical object keys, and shares the same cost as the length change for the really long keys.  The only savings the [s] approach yields in is in the case of single character keys.

Now @thebuzzmedia has already said he doesn't want to bring back the lower case character types with their special length handling, and I agree with the reasoning on that one.

However, here's some supporting evidence why [s] won't help with space savings on this proposal.

Look at these object keys from a size perspective:
<pre>
empty:
  [U][0] - existing
  [s][0] - [s]
  [0] - length change

short:
  [U][3][foo] - existing 
  [s][3][foo] - [s] 
  [3][foo] - length change

single char:
  [U][1][a] - existing
  [C][a] - [s]
  [1][a] - length change

long:
  [ I ][874] - existing
  [ S ][I][874] - [s]
  [255][I][874] - length change
</pre>


Using the [s] approach:
* the same as existing in the empty and short key cases (+0, +0)
* gains a byte in the single character key case (+3)
* loses a byte in the long key cases ( -1)

Overall out of 10 (+2)

Redefining the length:
* gains a byte gain in the empty, short, and single character key cases (+1, +5, +3)
* loses a byte in the long key case (-1)

Overall out of 10 (+8)

The same benefit of taking a byte out of every key in every object behind removing [S] from the object keys can be had a second time (with the minor caveat on super long keys).  Outside the object keys the gains are equally beneficial, and the downsides remain minimal.

The old encoder's output cannot be saved, the space savings is at least as big as taking [S] out of [{], and the code changes to the encoders/decoders is actually quite minimal if they've written a getCount(fileHandle) utility function that returns a 64-bit int (and if they haven't, this change is a good opportunity to do so).

I'm glad the empty string oversight was caught, and to have thoroughly and properly considered the [s]/[h] option, which would have really been the only other approach to yield similar benefits.

As far as Draft 12 goes, my thoughts are yes, definitely go for it.  I'd almost consider it a crime not to at this point. ;)
I recognize other's hesitations, but is anyone else up for seconding this?

**Comment by kxepal** [comment]
@MikeFair 
I didn't propose `[s]` to provide a lot of space saving. Just to handle 255 sized strings and HIDEFs which are common ones and cleanup the spec. If you really want to save significant amount of space, you better wrap your UBJSON with gzip/snappy instead of playing with the bytes (: 

If I would really introduce an additional magic byte for strings I would make it a reference to specific compression type which we could support. That's how [Erlang's binary terms](http://www.erlang.org/doc/apps/erts/erl_ext_dist.html) works, thought they have just a flag if data is zlib compressed or not. This will save much more space with a reason for having magic `[255]` byte. For now there is no much to have it.

**Comment by thebuzzmedia** [comment]
@kxepal 

> Since [C] wasn't removed, it's unclear way it couldn't be used for object keys. Now this problem is solved.

We had the 'can C be object key?' discussion a year or so ago - I would rather not open this discussion up again. Object keys are Strings.

> This may attract at the first sight, but such kind of optimization doesn't gives much space saving since container length defines only once. I don't think optimization at this place plays any significant role.

Absolutely agree - it's for consistency. We are discussing redefining _length_, so I gave examples of all the cases where this would change. Container length is one of those places (but as you stated, not a huge space savings).

> Introducing a new [s] and [h] applies transparently without parser change since all what you need is to add new tags handlers.

I understand your point, but I don't like how `s` and `h` now become special case versions of `S` and `H` - there is no matching short length definition for other types (containers). It feels... incomplete to me.

I think we are arguing aesthetics here, which is a lot like arguing religion :)

> proposed implementation while optimizes 90% of cases, introduces unwanted overhead for the other 10%

Yes, exactly, with the additional clarification that in the "10% case", the data being parsed is guaranteed to be big enough that the addition 1-byte overhead is trivially small.

> it accidentally causes specification cleanup in order to remove implicit type tags from object keys with no loss in comparison with @MikeFair proposal and is able to make one-char keys legal which allows to save one more byte for them.

No no no :) -- I don't want to go down this path. I think that is a rare enough case, introducing a weird enough change that I don't want to consider this. Object keys are strings.

The second we go down this path, the gentleman that asked us to allow Object keys to be ANY type value (from ~1 year ago) will come back and demand that Object keys are also numeric values and I will have to tell him "Will since I gave Alex CHAR, I'll give you NUMERIC!" and then all our parsers will be super confusing and we will have to quit our full time jobs just to understand UBJSON spec... maybe I exaggerated a little bit :)

> Simplicity is a key feature of UBJSON and branches for magic bytes were never made things simple.

This is a *good* point. I have seen the '255, read next char' design in a number of specs. If it weren't a relatively common thing to see, I would definitely veto it for this reason, but I think the win will be seen across most all of the payloads and it's a design that for other format designers, has been used time and time again before. Put another way, "common enough to not scare anyone".

That's my thinking at least.

**Comment by Miosss** [comment]
@thebuzzmedia
> I have seen the '255, read next char' design in a number of specs

Do you have some examples, please?

Gentlemen, one thing. Do not forget that we have multiple optimizations pending for approval. We should not argument our thinking here despite other proposals.
Specifically, 1 byte savings of [255], while now give savings for every string, may be of much less imporatnce, when discussing arays of objects, whose keys are predefined using typespec. In such case, we do not gaiun one byte, per key, per each array antry, but only per each key specified once (even if we have 10000 of elemnts in the array following, but of the same structure).
I just want you to remember, what may this particular issue mean when other proposals are agreed on.

**Comment by MikeFair** [comment]
> If you really want to save significant amount of space, you better wrap your UBJSON with gzip/snappy instead of playing with the bytes (:
> If I would really introduce an additional magic byte for strings I would make it a reference to specific compression type which we could support.

I totally agree, but realizing those savings requires really long strings with repetitive text or limited character usage to begin with, and is a different proposal that "fixes" the failings of this one. :)

Both recommendations stand on their own.

**Comment by MikeFair** [comment]
@Miosss 
>> I have seen the '255, read next char' design in a number of specs
> Do you have some examples, please?

The telnet protocol uses this technique for sending commands; 255 means "the next byte is a command" and if the value 255 is required then it's sent as 255 255.

> Specifically, 1 byte savings of [255], while now give savings for every string, may be of much less imporatnce, when discussing arays of objects

Arrays of objects optimizations are likely to do something for the object keys, but not for the string values themselves.  This is an opportunity to take a byte out in many simple places and while it's likely hard to predict exactly how much the savings will be I can't foresee a case where it would have been a mistake to be using this style of count.

I even think it makes looking at most examples that much easier to type and to read.
Just because a further optimization down the road might diminish the benefits seen by this one doesn't mean this one was somehow not beneficial.

And when talking about encoder/decoder update requirements, those optimizations will require way more work than this one does.

**Comment by breese** [comment]
I do not feel comfortable with the above proposals because they are too context-sensitive to me.

A better approach is to add small-integer tokens (similar to that MsgPack does.) For this to be compatible with the existing tokens, we need to transpose these tokens to another range. So we could use [128] = value 0, [129] = value 1, ... [255] = value 127. In other words, [128] is equivalent to [i][0].

Small-integers are not restricted to lengths, but can be used in any place where integers are needed.

**Comment by jnguiavarch** [comment]
For me, introducing `[s]` as a shorthand to `[S][U]` somehow breaks the TLV because L is then defined in two different ways depending on T.
That said I think @MikeFair's proposal of changing the way L is defined is a good idea. However, I would go a little bit further by not using the recursive definition of L at all. I would use some very basic variable length integer encoding. Say: The higher bit of each byte is 1 if there are more byte(s) and 0 for the last one, the other 7 bits encode the length itself. Reading the length would be very simple:
```c
length = 0
do {
  byte = bytes[position++]
  length = (length << 7) | (byte & 0x7F)
} while (byte & 0x80)
```
Using this the zero length string would be `[S][0]`, any string with length <= 128 would use 1 byte length, any string <= 16383 would use 2 byte length, ...
For me it would not change the specification of `[#]` in optimised containers because it's not a length it's a count. The length is the number of bytes of the payload.

**Comment by kxepal** [comment]
> For me, introducing [s] as a shorthand to [S][U] somehow breaks the TLV because L is then defined in two different ways depending on T.

It actually doesn't since in UBJSON L and V are optional and actually for `[s]` length is defined in the same way as `[S]` is defined for object keys - by implicit assumption.

**Comment by MikeFair** [comment]
@jnguiavarch 

> However, I would go a little bit further by not using the recursive definition of L at all.

I agree with you in principle; it's just so far turned out that nothing except for assuming the initial [U], actually creates a better performing solution (opposed to just being philosophically different and perform the same or worse). 


> The higher bit of each byte is 1 if there are more byte(s) and 0 for the last one, the other 7 bits encode the length itself.

I too keep wanting to have/use a format like this, but I discarded the idea of using any arbitrary length formats like this for a length because they simply don't perform well in comparison and the use cases they might outperform simply don't exist.  It's kind of counter intuitive, so I'll explain.

The actual point of getting the length in this use case is to instruct the CPU on how many bytes to suck up or skip.

The Achilles heels of the format are in the position++ and the shift <<7.  

The position++ operation ultimately has to operate on a CPU instruction aligned offset from some starting address. That means the "position" integer must be a CPU aligned data value and it will be smaller than the addressable memory of the architecture. 

This means that "position" can't, for any meaningful work, exceed a 64-bit value; first because we have no data items big enough; and second because there are no machines that are using/addressing more than 64-bits such that they could have encoded the value and transmitted the data in the first place (even our existing 64-bit machines are only putting in the hardware lines to address 48-bits of actual memory).
This is a form of "proof" that "position" can't ever be greater than 64-bits. Therefore, there is no length that can possibly be encoded that will be longer than that.  Therefore 64-bits creates a meaningful upper boundary on the size of a length; 64-bits/7-bits = 9.12 which means 10 bytes to store a 64-bit value in the proposed format.  

That's the same size of the [255][type][value] format.  The proposed format does provide a more linear growth to the byte usage, however once you pass using 2 bytes of length value and are processing more than 32K or 64K bytes of data, discussing the +/- 2-5 bytes is pointless.  

The worst case for the [U] first format is the 256 bytes case.  The difference between it and the continuation bit proposal is 260 bytes vs 258 bytes.  I think those are effectively the same numbers.

What's not the same number is that the proposed format is inherently slower to process for those two bytes in savings.  By putting a bit split transformation in the middle of the data, it makes the algorithm, in practice, significantly more expensive to run.  

That shift / or operation is saying "move every bit I've loaded so far over 7 bits, then merge in these new 7 bits at the end (and move everything 7 places again every next byte)".  

That ends up being a bit shift operation for every byte of value; in contrast the [type][value] takes the byte, branches to the code that executes the appropriately sized fetch and is done.  

The problem here is that the 1 bit of "metadata" or "processing information" is being mixed into the data values creating a non-CPU aligned boundary which ends up costing a lot of "performance" to separate it out from the rest of the value. To get a huge speed increase, instead of embedding the bits inside the data values, use a 1-byte header, and let those 8 bits be the continuation bits indicating another byte.  It wastes only 1-bit per unused byte, enables the full 8 bits in the data to be used, and comes to the decoder pre-byte aligned making it extremely fast to process.  However, it gives up its size advantage, and is effectively identical to using [type][value]; making the benefits of switching moot.


For instance, in the case of the format you put forth, if you used a single header byte instead of embedding the continuation bit, and then used the 8 bits in the header as a bit field to indicate continuation, then you wouldn't have to pay the split and shift costs when processing the data stream.  You'd "waste" 1 bit per unused byte and in exchange you'd get a byte aligned data stream that the CPU can directly operate on or could at least easily pad with bytes to make it aligned.

This is how I got to what I proposed actually.
Then once I did that, I recognized "hey, since I'm paying for that header byte anyway why not just make it a count of bytes.  Then it can represent up to 255 bytes of length for that one byte instead of just 8.  Further, the decoder will then know exactly how many bytes to suck up!  But it does sacrifice the ability to provide lengths represented by integers longer than 255 bytes. But really, who cares, 8 is already a totally insane number of bytes for a length anyway!"

Then I recognized, there is no meaningful difference between the [count][value] format and the existing [type][value] format; they are identical; the only difference is that [type][value] provides more flexibility for picking other arbitrary formats that might come along in the future where [count][value] is restricting itself to only using up to 255 byte long integers and must consider transforming its values into CPU instruction aligned values, whereas [type][value] most likely comes in already CPU instruction sized aligned.  

Both formats are more efficient to process than the integrated continuation bit format which makes a difference when the algorithm is being called frequently. But [type][value] gets the edge over [count][value] because of its future proofing protection.  And then coincidentally it just happens to already be the format UBJ is already using so there was nothing further to recommend for making a change other than "assuming there's a [U] first, because the fact is, there most likely is".

**Comment by MikeFair** [comment]
@breese 

> A better approach is to add small-integer tokens (similar to that MsgPack does.) 

I tried reading up on it but I didn't follow it; could you explain some more on how you see it?

> For this to be compatible with the existing tokens, we need to transpose these tokens to another range

The proposal is explicitly incompatible with existing generators/tokens.  It is creating a format explicitly for use as a count or a length in UBJ, optimized for values < 255 where the majority of lengths/counts live.  For lengths/counts >=255 it takes on a single byte of overhead, and claims that at those sizes and CPU processing requirements, the extra byte and if instruction is hardly anything noticeable.

**Comment by breese** [comment]
@MikeFair Recall that we have tokens that contains both the type and the value, such as the boolean types [T] and [F]. Small-integers belong to this category: the token with value 128 encodes integer value 0, token 129 encodes integer value 1, and so on all the way up to token 255 which encodes integer value 127.

Small-integers can be used in any place where we can use integers ([i], [I], [l], [L], and [U].) More specifically, they can be used as length parameter in type-length-value fields. This obliviates the need for special context-sensitive encodings as suggested elsewhere in this discussion.

**Comment by MikeFair** [comment]
> the token with value 128 encodes integer value 0, token 129 encodes
integer value 1, and so on all the way up to token 255 which encodes
integer value 127.

That's a really intriguing idea; use the "forbidden type values" as a set
of short ints (if the high bit is set on the [type], it's a short int).

I was planning on proposing to use that range for runtime dynamic/user
registered types instead (the defined spec uses the positive values; while
the at-runtime types use the "high bit set" values); are there any other
plans for using that range?  All I read was that the range is excluded from
being a type, but there was no explanation I saw as to why they were
excluded.

If there was something else to see all those values get used for all at
once; this would probably be it.  I don't have the same attachment to
preserving the existing [type][value] length definition, and I see this
idea is just as context sensitive as the other; it's just defining it the
other way around and with a bit instead of a byte; or using 127 special
values instead of just one.  I also think defining length as its own thing
is better long term than collapsing it as being an int TLV thing; but
that's more a philosophical thing I think.

The only really negative point I see is limiting the 1 byte string length
opportunity to 127 instead of 254.  Which excludes things exactly 128 long
(and longer obviously) like UUID, GUID, unshortened IPv6, etc. I saw
including strings up to the 180-200 length range makes the most practical
use out of the 1-byte length.

So the question is what are the more important things to preserve and build
for? Is 127 "big enough" for a length and does this seem the best use of
the type values 128-255?  In which case we use them; this remains totally
consistent will existing generators (addressing @thebuzzmedia's very first
compatibility question) and I think hits many of the markers of the
existing spec as it's been defined so far.   This breaks the "type values
over 127 are forbidden" rule and preserves pretty much everything else I
can see. It doesn't burn a byte for longer values and instead creates a new
set of "short int" types that can be used anywhere a [U]<=127 is called
for, including a length (and unlike the 's' proposal, this idea remains
completely compatible with the existing descriptors for an object's keys).
If 127 is "long enough", then this is clearly the better option.

If 128-254 values are important as 1-byte lengths (and I personally think
they are because common 128 length types of values are known; but I can see
excluding them), then it requires an actual full byte and redefining the
"forbidden type values" as no longer being forbidden values won't work.

Other than preserving existing object formats (which breaks under the [U]
byte first proposal), there's nothing that excludes doing both.  Change
length because of the range it offers; and bring in a "short int" type
because of the byte it saves on all int values <128.  This breaks
recursively declaring an L as a TLV.

I can see it going either way. I think of a length as its own thing
separate from a TLV; one can use a TLV to define a length, but a length is
not necessarily a TLV. A decoder should have a function called getLength();
to use when a length is called for.  That doesn't automatically nake the
[U] byte first proposal better; it just makes the argument "but it breaks
using TLV for length" less meaningful and shifts the arguments to other
focus points.

**Comment by Miosss** [comment]
@breese @MikeFair 
Messing up with bit-level optimizations may be tempting, but has performance penalties - spliting bytes into portions cost another CPU instructions, and we want UBJ to remain as fast as possible (for transparent JSON conversion).

Using [255] as proposed by @MikeFair is the way to go. 
From benchmarks I made (I will show them to you, when I am more finished than now :) ), the situation is as follows:
- keys in objects are ALWAYS shorter than 255 bytes (this means like 99,999%; I just didn't found any key longer, while testing around 10kk of keys - 400MB of JSONs).
- string (other than keys) longer than 255 are less then 5% (usually of course)
- average string/key length is usually below 10 bytes, mostly about 6-7 bytes

This proposal offers 1B savings for each string shorter than 255.
What's more, because of the average string length, this proposition offers around 10-20% optimization for most strings. In 100MB JSON files I examined, it gives aroung 4MB savings.
For string longer than 254, it performs worse - it adds 1B. But it is less than 0,4% overhead, which is still very good. And those strings are rare, and can be much longer, for example 1024 bytes (it becomes around 0,01% overhead).

I think it is the way to go for simple, yet effective strings optimization.

**Comment by breese** [comment]
@Miosss I do not know what kind of performance penalties you are referring to -- the fact that they have the most significant bit set is purely accidental. Small integers are literal integer values, just like [T] and [F] are literal boolean values. They have the exact same performance overhead as boolean values.

**Comment by Miosss** [comment]
@breese 
I maybe misread your proposal - I thought you referred to splitting the byte of type - where some part would mean type and second part - length.

If you want ale type-bytes >= 128 to mean string of some length - then this is huge breaking of TLV.

By the way, we refer to the TLV so often, but V does not come always, and L only in strings...

**Comment by breese** [comment]
@Miosss And you are still misreading it :-) My proposal does not break TLV. You have to read my proposal in two steps.

First -- and here you have to forget all about lengths -- I am proposing a new kind of integers in addition to those that we already have. These small integers can be used in any place where we can use integers todays.

Second -- and now we bring back lengths -- because lengths are encoded by integers, they can also be encoded by small integers.

So a string currently encoded as
```
[S][i][5][hello]
```
can be encoded with small integers as
```
[S][i5][hello]
```
where [i5] is the small integer 5, whose encoding is 133 (= 128+5.)

**Comment by Miosss** [comment]
@breese 
Oh, I get it now.

But here we have less than with the [255]. I thought that you speak of something like:
```
[S5][hello]
```

**Comment by MikeFair** [comment]
@Miosss

> But here we have less than with the [255].

Yes, the limit using short int would be [127] (by consuming the type values
128-255 and subtracting 128) which ought to be big enough for 90%+ of
string lengths; and it preserves the existing practice of treating L as
just an int kind of TLV (keeping compatibility with the existing format
for object keys and preserving the length is a TLV definition).  By
excluding the 128-255 range of values it doesn't have the "extra byte"
perceived as extra/wasted space (though it starts requiring the extra byte
for anything in the 128-254 range).

Overall it's a really elegant solution and if it didn't have the 127
limitation, I'd be in totally favor of it (I'm not strongly against it
either; however I can't see it as clearly superior to the 254 solution). To
be clear, it's trading one kind of special number for length for another.
Rather than redefine length, this creates 127 new value inclusive types of
short ints (it's still testing the value of that first byte to determine
what to do next, it shortens it to do this once instead of twice for the
larger values). It's a bit odd to consider this creates 127 types which
covers half the value space of [U].

A similar thing to consider would be to define special values of Length:
253=[I], 254=[l], 255=[L]
This would be the equivalent solution to get rid of that "extra byte" for
values >252.  I'm not recommending this, simply using it to illustrate how
all methods are using "special values" in that first byte to mean something.

I have five comments on it:
1) outside use as a length I imagine ints <=127 are very common, but are
they so much more common than the other half (128-255) to warrant dedicated
type values?  I don't know...  It seems values 32 and under would capture
most that range as a length.

2) This consumes the 128-255 values as types. Values that are explicitly
excluded currently; what was/is the reasoning for that exclusion? Does it
apply to this use case? (defining it as >=128 does make it easy to detect
without specifying an encoding.)

3) That range could be used for a more dynamic, user defined, type of use.
While short ints seem a good thing to have, consider that once that range
is allocated for this use it's not coming back.  Good part of this is that
the current spec excludes that range, so it isn't being used today and this
gives it a use (which could be good depending on the answer to question 2).
The part to consider is this consumes that whole range in one swoop for int
values 0-127, is there no other future use of these values worth preserving
the range for?

4) It's not getting the full range up to 254 and specifically stopping
short of capturing the 128 length strings (the longest well known "short
strings").  How common are the 128+ lengths? They seem uncommon enough to
be an acceptable exclusion, but I'd prefer to see them included...

5) Using the 255 approach identifies/treats a length as a separate thing
from a TLV.  Breaking the mental association between a length and a TLV
allows future modifications to either a length or TLV be evaluated
independently of their impact on each other.  For example, if it ever
became a good idea to make [type] two bytes instead one, then regardless of
the benefits to the type definition, the impact on length would need to be
considered.  Separating them from each other now makes it clear that a TLV
is a type, a length, and a value; and a length is a length not a TLV
(though it might use a TLV as part of its definition).  Changes to one does
not necessarily mean a change to the other.

Short ints could be seen as reinforcing the link between length and TLV.
If the link is a good thing, then this reinforcement is good; if not then
short ints and lengths are separate conversations.

**Comment by lifthrasiir** [comment]
Another way to increase the limit is to swap the literal integer range and the type tag range: [i], for example, will encode to `E9` instead of `69`. Since ASCII's C0 range (remapped to `80` through `9F`) is almost unused, we can repurpose it as an additional integer range and extend the limit up to 160.

**Comment by MikeFair** [comment]
> Another way to increase the limit is to swap the literal integer range
and the type tag range: [i], for example, will encode to E9 instead of 69.
Since ASCII's C0 range (remapped to 80 through 9F) is almost unused, we can
repurpose it as an additional integer range and extend the limit up to 160.

This gets up close enough to the ~180 target (128 plus some "other useful
stuff") to eliminate the range concern I had; though I think this goes
beyond UBJ's intent (the [i] would no longer read 'i' when looking at the
hexdump which I think would break UBJ's commitment to being 'simple'); and
it doesn't share the benefit of keeping the existing object key lengths
intact, which I see as the primary strength of the 128+ proposal.

That first byte, regardless of the method chosen, is going to have a
mixed-use interpretation.

The 128+ approach completely preserves the existing TLV format and captures
most of the short length benefits; which obviously makes it worth
considering.  If keeping length a TLV is important, then this would be the
route to go.  (It's even worth having a short int on its own independent of
using it as a length.)

I've already made my arguments for separating count (or length) from a TLV;
making a count that's optimized for representing an integer; and that
distinguishing the two is best for the long term.  I haven't heard/seen
much comment arguing for/against keeping them the same or separating them.

The 128+ approach keeps TLV and won't break exisitng generators (i.e. Newer
parsers can still read in the older format).

The 255 approach brings with it a seeming suboptimization in cost for
longer lengths. Analysis can show this "cost" as practically nothing and
hardly a negative.  The bigger consideration is this is a diverging change
for length; it's separating length/count from a TLV.

And lastly, we've got the idea of using the top few ints of the range
(252-255) to represent the different integer types.  Which eliminates that
1-byte "cost" aspect of the 255 approach though I think introduces more
complexity/rigidity than is desireable.

If collapsing a length as a TLV is a good thing, then I'd recommend the
128+ solution.

I think it's better though to define a length as a 64-bit thing where the
value is encoded on the wire as a length.  This hardcodes UBJ to be
restricted to 64-bit sized things.  Which means the functions to read/write
a length in the encoders/decoders will return/take a 64-bit value.  And
that will make them simpler to write/use and should make the overall
decoding process faster because all counts/lengths will be of a known size
(no code logic needed to figure it out and use different functions every
time).  For most scripting languages this is likely a non-issue; but for
the strongly typed languages it should make things simpler.

**Comment by MikeFair** [comment]
@thebuzzmedia

can you clarify why 128+ was/is excluded from being used as a type value?

It would have bearing on whether or not using 128+ as a short int type is feasible.

**Comment by moridinga** [comment]
You guys are just recreating the wheel, a la ASN.1 BER definite length octets: http://en.wikipedia.org/wiki/X.690#The_definite_form

Just follow that syntax and be done with it. You get the single byte optimization for the common case of short strings (up to 127 bytes), and you get a compatible length definition that can extend to 127 bytes. And the selection between the two is the top-most bit.

**Comment by lightmare** [comment]
Hi, I've just stumbled upon this discussion, and am wondering whether you've come to a conclusion (somewhere else perhaps).

Regarding *small integers*: If you're specifically worried about 128-byte long strings, you could use ASCII digits `'0'` .. `'9'` as type markers for the first 10 small integers (just like `'T'` means `boolean true`, `'5'` would mean `integer 5`), and then types 128 .. 255 would map to integers 10 .. 137. But although I really like the idea, I wouldn't vote for small integers 128+. One reason being that they eat the whole "forbidden but possibly usable for user-defined extensions" space.

I think that `<LENGTH>` should be a completely separate kind of entity. The fact that it's currently encoded as a valid `TLV` construct doesn't make much sense to me. There are constraints on what it can contain (non-negative integers), anyway.

I like the OP proposal, and would take it further by encoding the `<TYPE>` in the first byte. There are only 4 allowed types, and you may reduce that to 3 if you don't mind values `253` .. `255` will be encoded as `int16` instead of `uint8`.

Note: I'll use `<J>` for `int16` and `<K>` for `int32` because it's difficult to distinguish between uppercase `<i>` and lowercase `<L>`.

**`[255] TLV`** -- OP proposal, value 255 is special
```
(< U > != 255) ? < U > : < TYPE >< VALUE >

252 => uint8
253 => uint8
500 => [255][J] + int16
50000 => [255][K] + int32
5<<20 => [255][K] + int32
5<<40 => [255][L] + int64
```

**`[253+x] (16 << x)-bit uint`** -- values 253 .. 255 are special
```
(<U> <= 252) ? <U> :
(<U> == 253) ? <unsigned J> :
(<U> == 254) ? <unsigned K> : <unsigned L>

252 => uint8
253 => [253] + uint16 // +2B
500 => [253] + uint16 // -1B
50000 => [253] + uint16 // -3B
5<<20 => [254] + uint32 // -1B
5<<40 => [255] + uint64 // -1B
```

**`[252+x] (8 << x)-bit uint`** -- values 252 .. 255 are special
```
(<U> <= 251) ? <U> :
(<U> == 252) ? <next U> :
(<U> == 253) ? <unsigned J> :
(<U> == 254) ? <unsigned K> : <unsigned L>

252 => [252] + uint8 // +1B
253 => [252] + uint8 // +1B
500 => [253] + uint16 // -1B
50000 => [253] + uint16 // -3B
5<<20 => [254] + uint32 // -1B
5<<40 => [255] + uint64 // -1B
```

**Comment by MikeFair** [comment]
@moridinga 

> You guys are just recreating the wheel, a la ASN.1 BER definite length octets:
>  http://en.wikipedia.org/wiki/X.690#The_definite_form
> 
> Just follow that syntax and be done with it. 

Unfortunately, using the ASN format directly would mostly just combine the worst consequences from all the proposals.

Because it uses the values 0-127; it clobbers over the same value space as UBJ's existing types defeating the whole purpose of introducing the short int value types in the first place (which was that it would preserve backward compatibility with the existing spec); and it includes the 127 length limitation for the 1-byte usage case; stopping just short of that all too commonly encountered 128 length requirement.

The long form (everything >=128) is equivalent to the existing practice in terms of space (type byte followed by a value) except the ASN format ends up being more costly to parse (because of the possibility of non-CPU instruction aligned data values and values larger than the machine can directly support); if the byte count values were restricted to 2, 4, and 8, then it would be, for all meaningful comparisons, identical to the existing practice.

The only practical use cases for the long form are where the number of bytes <=6 (and most likely <=4) meaning that the value range it allows of byte counts 7-127 is a waste of a specification because those values can't ever actually get used in practice (those data amounts are simply too big to realistically transfer and address).

**Comment by MikeFair** [comment]
@lightmare

I agree, this would be a net good thing for the spec overall in the next draft.  
Technically the spec could just call all the values 248 and above "reserved" in this draft and only define a couple of them for now.  I was originally trying to limit the number of special values to exactly 2 (0 and 255) but now recognize eliminating the "extra byte" in all the large values is more valuable than the cost of adding a couple more "special values" where an existing "special value" already has to be.

I would define them backwards (or downwards) though to demonstrate a pattern for future growth if called for (despite the fact that 2^64 is just such incomprehensibly huge number already).

So the special cases right now would be:
254 - 16-Bit
253 - 32-Bit
252 - 64-Bit
0 - empty string

This practically eliminates the "extra byte" issue from the original proposal.  The only cases that get introduced, like mentioned, are lengths of value 252-254 which get pushed into being 16-bit things instead of the 8-bit things they might have been.  Losing a byte in those 3 specific cases is arguably better than losing a byte in all of the large cases, so I'm totally good with that.  

I'm even leaving the 255 value open so it can still be used as part of the 1-byte length (because 255 is a byte boundary; it's more likely to appear than other nearby values).  But that's easy enough to give up if it really rubs people the wrong way.  

We'll see what @thebuzzmedia has to say.

**Comment by lightmare** [comment]
If values >= 248 are reserved, all of them still need to be encodable as length.
Either as ` [254] <uint16>`, or use another special case for `[25x] <uint8>` (where that uint8 will be >= 248, but encoding 3 bits of information into uint8 is still better than uint16 ;))

**Comment by MikeFair** [comment]
@lightmare

Agreed, they are all encoded as 254 (16 bit); with the possible exception
of 255 (which becomes the psuedoexception special case value that is
actually encoding the length 255).

Adding a special 25x 8bit value would only have meaning to provide lengths
for these specific values and would only save a single byte. Consuming a
special value to handle at the most 8 specific, not particularly common,
cases seems an aweful waste to me.

At present 256+ are already 16-bit values; this is just lowering that
threshhold to 248+.  Then creating a possible exception opportunity for the
value 255 because as a power of 2, byte boundary value, it might be a bit
more common.

To be clear; I think 255 is a way of declaring "the largest array indexable
by a single byte"; and consequently gets used to preallocate space in data
structures because of that; and therefore it becomes a length value to
consider preserving in the set of single byte values.  On the contrary,
252, 253, and 254 aren't likewise special for any particular reason and so
make for better use as special cases and don't warrant preserving as a
single byte value.

Thoughts?

**Comment by lightmare** [comment]
**Problem**
LENGTH 125 == ASCII '}' so inside an object we can't tell whether the object is ending or a 125-byte key follows.

For example -- 1 object containing 1 key-value pair:
```JSON
{ "S|This.is.a.string.that.starts.with.S.PIPE.and.is.125.bytes.long...": "puzzles" }
```
```
[{]
  [125] [S|This.is.a.string.that.starts.with.S.PIPE.and.is.125.bytes.long...]
  [S] [7] [puzzles]
[}]
```

is binary equivalent to -- 1 empty object, 2 strings and a dangling `}`:
```
[{] [}]
[S] [124] [This.is.a.string.that.starts.with.S.PIPE.and.is.125.bytes.long...]
[S] [7] [puzzles]
[}]
```

**Solution A**
When used as object key length, 125 must be encoded in 3 bytes as `[254] <uint16>`.

**Solution B**
Change the object-end marker to a special value >= 248.

**Solution C**
Escape the closing `}`. Either with a distinct special value, or with a silly-encoded length, like `[254] [0]`. With the silly-encoded zero length approach, if the object actually contains a zero-length key, then after that key was read, the closing escape doesn't have to use 3-byte silly encoding, but simply `[0]` because zero-length key shouldn't appear twice. I think this solution is silly, putting it here for completeness. Its only benefit is that it keeps `{` `}` balanced.

**Comment by lightmare** [comment]
ad **Solution C**
Should've thought that out better. Simply `[254] [125]` could be **the** object-end marker sequence. It's a silly-encoded length=125, and visually it's "some byte, `}`, zero byte".

**Comment by MikeFair** [comment]
> Problem
> LENGTH 125 == ASCII '}' so inside an object we can't tell whether the
object is ending or a 125-byte key follows.

Great catch;

I'll throw in Solution D, put field/value pairs into counted sets.

This would mean putting a length before a set of field/value pairs.

Typically the number of fields for the whole object would be known up front
and go immediately after the { (just before the pairs). However, if that's
not known then the encoder can put the quantity of pairs it does know.

(Perhaps consider rotating the object to be two arrays; an array of field
names followed by an array of values to line up with the CTLV format? ).

**Comment by MikeFair** [comment]
Actually, I think you're on to something with Solution C.

I see this as encoding "terminating a list of lengths".  And I don't think this is the only place/time this kind of ambiguity is going to exist.  Allowing a terminator type character like 125=}, 93=], or other type indicator where a < length > is supposed to be is ambiguous.

I'm thinking about what the code for the parser of an object would look like here.  A main point aside from size is also speed by eliminating extra reads/branches and giving the parser certainty.  

I've been advocating for parsers to write a function called getLength() to retrieve a length as an unsigned 64-bit value whenever a length is called for. If I change this to a signed 64-bit int, then -1 can be used as a return code for an invalid length (or some other function/error/exception handler).

Generalizing Solution C, the value [254] would simply become an "invalid length" to force error/exception/escape conditions which gives parsers a way to deal with this.

    int64 getLength() {  // return a signed 64-bit int
        uint8 b = getByte();
        switch (b) {
            case 255:
                return (int64)255;  // return the value 255
            case 254:
                return -1; // return error code -1 -- Or throw an exception for languages that have it
            case 253:
                return (int64)getUInt16B(); // unsigned 16-bit Big endian
            case 252:
                return (int64)getUInt16L(); // unsigned 16-bit Little endian
            case 251:
                return (int64)getUInt32B(); // unsigned 32-bit Big 
            case 250:
                return (int64)getUInt32L(); // unsigned 32-bit Little 
            case 249:
                return (int64)getInt64B(); // signed 64-bit Big (signed to support the -1 return value)
            case 248:
                return (int64)getInt64L(); // signed 64-bit Little (signed to support the -1 return value)
            else:
                return (int64)b;
        }
    }

    JSONString getString() {
        int64 L = getLength();

        if (L == -1) {
            return 0; // or null; for languages that support null -- basically an "invalid string"
        } else {
            return new JSONString(readStringBytes(L));
        }
    }

    JSONObject getObject() {    
        JSONObject obj = new JSONObject();
        JSONString k;

        int done = 0;
        while (!done) {
            k = getString();
            if (k != 0) {
                v = getValue();
                obj.add(k, v);
            } else {
                done = 1;
            }
        }
        char c = getByte();
        if (c != '}') { 
            // Some error occurred or other command codes/types can be defined here
        }
        return obj;
    }

There's probably better ways to handle this overall, but this code is using a null string value for a key (i.e. stringPtr == 0) as the termination condition.  It worked with the procedures getObject(), getString(), getValue(), and getLength() and did not require anything to "peek" at the next value to figure out what to do or need any context about what object it was currently dealing with.

This is different semantically from Solution C although it's using the same encoding ( [254][125] ).  

This approach is saying anytime you need to terminate a list where lengths are the expected value; use the value [254]. Which is kind of the same thing as treating length [254] as an escape code.  

I can think of other places where this might be useful. Like the indeces for nested arrays of arrays -- the list of indeces can now be < lengths > like they should be, with [254][#][L][[] starting a nested sub-array; and for the arrays of objects where the < length > would specify how many objects follow in the next block, and each section can either continue to adding more blocks of objects or terminate the array.
